DESIGN.md

```diff
@@ -0,0 +1,111 @@
+
+The goal of this project is to add schema change functionality to Terrier, and do it in a non-blocking, lazy way. The idea is to store a separate Datatable for each schema version, and do it Our goals are:
+
+75%: Support lazy add/drop column and arbitrary number of schema versions      (mostly done, need to update the upper level of the system and the catalog, and pass benchmarks)
+
+100%: Do background migration of tupleslots. Support GC of old schema versions and datatables. 
+
+125%: Support unsafe schema changes, which includes updating type, nullable, or default value status of columns.
+
+150%: Codegen layer to precompile tuple transformations.
+
+We have yet to design our system for unsafe schema changes and precompilation. In this design doc we will focus on the 75% and 100% goals.
+
+## Scope
+
+Our project will mainly modify Sqltable and Catalog. Sqltable will be modified to handle multiple datatables . The datatables are transparent to the users of Sqltable,  who view Sqltable as a logical table with multiple versions. All functions in the Sqltable API (such as insert/delete/update/select/scan/begin/end/projectmapforoids) are augmented with an extra argument, layout_version. layout_version tells Sqltable which version of the table the function is targeted to.
+
+We will modify the Catalog to support the functions DatabaseCatalog::UpdateSchema()/GetSchema()/GetConstraints(). The catalog will be store information about the schema versions of each Sqltable, and use the timestamp of a transaction to infer the correct schema version of each Sqltable the transaction accesses. 
+
+We will not change the API of Datatable, and make only minor additions to it. For convenience, within each Datatable we store its corresponding version, default value map, column id to oid map, and column oid to id map (explained below). Similarly, we will not be touching Tupleslots, Rawblocks, or Index.
+
+We will also make minor changes to upper levels of Terrier, such the execution layer (e.g. storage_interface.cpp). 
+
+## Architectural Design
+
+To manage multiple datatables, each Sqltable keeps a ordered map (tables_) from version number to a struct containing the metadata of each datatable (called DataTableVersion). 
+
+Note that in the vast majority of cases, there will be no schema changes, and each Sqltable manages only one Datatable. If we need to go through tables_, which is a concurrent map (or vector), on each call to Sqltable, that will be a huge overhead. Therefore, we will cache the DataTable in the SqlTable when there is only one DataTable to bypass the map/vector lookup.
```

Comment:
This appears not to be implemented. Is this a TODO or did something change

Reply:
yes it is a TODO